### PR TITLE
Fix #24091 - Selected configurable product attribute options are not displaying in wishlist page.

### LINKED ISSUE #22503: Added Wish-list product displaying with wrong configurations
### Preconditions (*)

1.Magento2.3 with sample data

### Steps to reproduce (*)

1.Log into the customer account.
2. Navigate to product listing page.
3.choose the configurable product and select preferred color and size options.
4.click on the wishlist icon of another configurable product on the same page without selecting any color/size options.

### Expected result (*)

1. On the Wishlist page, the system should not assign different product options to the wishlist product.

### Actual result (*)

1. system assigning wrong/other product configurations to the wishlist product.

qa-contribution

### LINKED ISSUE #24091: Selected configurable product attribute options are not displaying in wishlist page.
### Preconditions (*)

1. Magento 2.3.2 with sample data
2. Product listing page should have 2 or more configurable products with swatches.

### Steps to reproduce (*)

1. Log into the customer account.
2. Navigate to product listing page.
3. Choose the configurable product and select preferred color and size options of any product expect last product which is listed in Product listing page ( It is working fine for last product of the listing page).
4. Click on the wishlist icon of the product.

### Expected result (*)

1. On the Wishlist page, the selected configurable product options color and size should be visible. (See Details is available on the product, see this gif below
SeeDetails
)

### Actual result (*)

1. The selected color and size is not visible in wishlist page.

plp
wishlist

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
### Description (*)

This fixes #24091 and #22503
The reason of the issue is that in add-to-wishlist.js there are event observers that are being triggered on for every configurable product on page after modifying one of them. So if the last product on page doesn't contain any option checked, no matter which will be added to wishlist, it won't contain options.

### Fixed Issues (if relevant)

Fixes #22503
Fixes #24091 

### Manual testing scenarios (*)

Check fixed issues for steps to test

1) Created two Configurable Product 
2) The born products must be with one or more 'VisualSwatchOption', and related to one category.
3) Go to Category Page
4) Select 'VisualSwatchOption' for a first product in the list category and add to Wishlist
5) For added product must be present 'See Details' on Wishlist page

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Bartlomiejsz** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Unit Tests

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @Bartlomiejsz thanks for the pull request, it looks great! Looking at the problem, I believe we can solve this just on the JS part, follow below my thoughts:

Since the `_updateWishlistData` is triggered by an event, we have the `event.target` that is the selector of the option that the user has changed within the product block (it can be the product page or an item within a product list), from that perspective we can extract the selected attribute values from the correct product and also on `_updateAddToWishlistButton` we can update the post data on the right `add-to-wishlist` element. 

Using this approach, it won't be a need to have an object `dataToAdd` with productId, since the event gives us the context that we need to collect the data from the right product and update it's form.

We would still have to have some mechanism to figure out if we are on list or product page, but I think that can be done on JS as well, what do you think?

Please let me know if there are any questions.

**Comment by Bartlomiejsz** [comment]
Hi @gabrieldagama, that is great idea to limit execution of wishlist button update to only current product instead of lopping for every item possible. I managed to reduce js needed for it and also accomplish that, thank you.

However I'm not sure if checking if current page is product list via js is good idea.
This might cause some issues easier on modified themes with different structure/classes, and by assuming that we're checking i.e. body class - in the core there are already two possibilities (catalog-category-view, catalogsearch-result-index), and if somebody would like to add some custom products list page it would require to rewrite wishlist js just to make it work by adding additional class to check. With current solution this is only about adding one additional parameter to block.
Also, current solution is consistent with final price renderer, which is already part of the core.

Could you please check current solution after my changes and let me know what do you think?

File: app/code/Magento/Wishlist/view/frontend/layout/catalog_category_view.xml [hunk 1]
```diff
@@ -21,7 +21,15 @@
                        template="Magento_Wishlist::catalog/product/list/addto/wishlist.phtml"/>
             </referenceBlock>
             <referenceContainer name="category.product.list.additional">
-                <block class="Magento\Wishlist\Block\AddToWishlist" name="category.product.list.additional.wishlist_addto" template="Magento_Wishlist::addto.phtml" />
+                <block
+                    class="Magento\Wishlist\Block\AddToWishlist"
+                    name="category.product.list.additional.wishlist_addto"
+                    template="Magento_Wishlist::addto.phtml"
+                >
```
Comment (gabrieldagama) [inline app/code/Magento/Wishlist/view/frontend/layout/catalog_category_view.xml]:
```suggestion
                <block class="Magento\Wishlist\Block\AddToWishlist"
                    name="category.product.list.additional.wishlist_addto"
                    template="Magento_Wishlist::addto.phtml">
```

File: app/code/Magento/Wishlist/view/frontend/layout/catalogsearch_result_index.xml [hunk 2]
```diff
@@ -14,5 +14,17 @@
                        template="Magento_Wishlist::catalog/product/list/addto/wishlist.phtml"/>
             </referenceBlock>
         </referenceContainer>
+        <referenceBlock name="wishlist_page_head_components">
+            <block
+                class="Magento\Wishlist\Block\AddToWishlist"
+                name="catalogsearch.wishlist_addto"
+                template="Magento_Wishlist::addto.phtml"
+            >
```
Comment (gabrieldagama) [inline app/code/Magento/Wishlist/view/frontend/layout/catalogsearch_result_index.xml]:
```suggestion
            <block class="Magento\Wishlist\Block\AddToWishlist"
                name="catalogsearch.wishlist_addto"
                template="Magento_Wishlist::addto.phtml">
```

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @Bartlomiejsz, thanks for the updates! Agree with your arguments there, please follow some very small suggestions to keep consistency, also we must cover all PRs will tests, the `add-to-wishlist` has a jasmine test that may need to be updated based on the changes and it would be great if we can cover this with MFTF tests, let me know if you need any help with the tests!

Thanks!

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Hi @Bartlomiejsz. I'll take care of the test coverage if you don't mind.

**Comment by Bartlomiejsz** [comment]
@engcom-Echo i would not be able to do it this week so that's great, thank you!

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run Functional Tests CE

**Comment by gabrieldagama** [comment]
@magento run Functional Tests CE

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8062](https://jira.corp.magento.com/browse/ENGCOM-8062) has been created to process this Pull Request
:eight_spoked_asterisk: @gabrieldagama, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by Bartlomiejsz** [comment]
Actually, this seems to fix also https://github.com/magento/magento2/issues/22503

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Magento 2.4-develop with sample data
The product listing page should have 2 or more configurable products with swatches.

### Case 1

**Manual testing scenario:**

1. Log into the customer account.
2. Navigate to the product listing page.
3. Choose the configurable product and select preferred color and size options of any product expect last product which is listed in Product listing page ( It is working fine for last product of the listing page).
4. Click on the wishlist icon of the product.

**Before:** :heavy_multiplication_x: The selected color and size is not visible in the wishlist page.

Peek 2020-09-01 10-49

**After:** :heavy_check_mark: The selected color and size is visible in the wishlist page.

Peek 2020-09-01 11-25


### Case 2

**Manual testing scenario:**

1. Log into the customer account.
2. Navigate to product listing page.
3. choose the configurable product and select preferred color and size options.
4. click on the wishlist icon of another configurable product on the same page without selecting any color/size options.
For example:

qa-contribution


**After:** :heavy_check_mark:  the system didn't assign different product options to the wishlist product.

Peek 2020-09-01 11-28

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Integration Tests, Functional Tests CE

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by gabrieldagama** [comment]
@magento give me test instance with edition ee

**Comment by gabrieldagama** [comment]
@magento give me new instance with edition b2b

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

Magento 2.4-develop with sample data

### Case 1

**Manual testing scenario:**

1. Log into the customer account.
2. Navigate to the product listing page.
3. Choose the configurable product and select preferred color and size options of any product expect last product which is listed in Product listing page ( It is working fine for last product of the listing page).
4. Click on the wishlist icon of the product.

**Before:** :heavy_multiplication_x: The selected color and size is not visible in the wishlist page.

Peek 2020-10-15 15-08

**After:** :heavy_check_mark: The selected color and size is visible in the wishlist page.

Peek 2020-10-15 15-02

### Case 2

**Manual testing scenario:**

1. Log into the customer account.
2. Navigate to product listing page.
3. choose the configurable product and select preferred color and size options.
4. click on the wishlist icon of another configurable product on the same page without selecting any color/size options.
For example:

qa-contribution


**After:** :heavy_check_mark:  the system didn't assign different product options to the wishlist product.

Peek 2020-10-15 15-03

**Comment by gabrieldagama** [comment]
@magento run Functional Tests B2B,Functional Tests EE

**Pull request merged**